### PR TITLE
Annie ir ima vis

### PR DESCRIPTION
**Pull request by annierose3**
new IR IMA Vis notebook and all files that go with it.

### CODE REVIEW

**Review by FDauphin — CHANGES_REQUESTED** [review]
Nice first draft on the notebook. Visualizing IMAs has always been tricky so users will find this super helpful. Some overall comments:

- The docstrings should be in the Numpy convention (Parameters and Returns vs Inputs and Outputs). This keeps all documentation across notebooks consistent.
- Whenever possible, use descriptive variable names. It makes the code easier to read and a lot easier to maintain.
- In your functions which compute summary statistics, you loop through each image then compute. `numpy` functions usually have a variable called `axis` that essentially does that for you. Utilizing that will make your code more concise.
- There's a larger comment for better organizing the functionality of your diff and stats functions. TLDR you would gain A LOT from having one function for differencing and one function for summary statistics; not required but HIGHLY suggested.
- The names of your modules should be lowercase with underscores (i.e., `difference_methods.py` and `panel_show.py`). Furthermore, since the modules are small you could just have them under one module, but if you anticipate more differencing and panel functions to be implemented, then it's best to keep them separate.

I'll add the notebook review tomorrow; awesome work so far though 👍🏾

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:553]: is simply subtracting each read from the one before.

Also aren't you finding the median difference?

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:564]: It may be useful to define a function for these plots since you use them a few times.

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:635]: Again mean or median?

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:652]: Same comments for other plotting code blocks.

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:726]: Comment works better in prose as markdown text.

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:702]: Same comments as previous plotting code blocks.

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:718]: If you're changing all values above/below 3/-3, this might be a more condense way to do that:
```
cube[np.abs(cube) > 3] = np.nan
```
Also how is x being called?

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:836]: Hyperlink to the notebook subdirectory directly.

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:868]: Update with new link:
https://numpy.org/citing-numpy/

**Review by FDauphin — CHANGES_REQUESTED** [review]
The notebook is in really good shape. It's easy to follow and the plots are exceptional. Here are my overall notes:

- Put spaces when defining objects for readability e.g. `a = 5` not `a=5`.
- Use `path, filename = os.path.split(file)` for all plots to save a line
- Add whether the observation is either scattered or nominal to the subplot title (it'll just be more clear which is which)
- Try to make functions for the plotting. I think a lot of the code blocks could be condensed, which would make for a more readable notebook.
- Choose more appropriate variable names instead of one letter variables so the code blocks are easier to follow.
-  calwf3 should be `calwf3`

Solid work so far, I hope these edits go a long way 🙌🏾

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (FDauphin) [inline notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:641]: Why does the difference decrease with time again?

File: notebooks/ir_ima_visualization/PanelShow.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/PanelShow.py:80]: The title always shows the nanmedian of the cumulative difference, regardless of differencing method choice. Should it be the nanmedian of instantaneous difference? If not, that should be explicitly stated to not confuse users.

File: notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb
Comment (annierose3) [reply notebooks/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.ipynb:438]: electrons, in this plot. Later plots are in electrons/second.

**Comment by FDauphin** [comment]
Closing since everything is now in #43

**Comment by FDauphin** [comment]
Reopening to see if I can have all the commits from #43 pushed to here.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:10]: Requires the path to a RAW full-frame IR image fits file (filename).

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:73]: `t_0, t_j, and t_j_1` variables may also help with readability. I apologize, I thought that was assumed from my comment.

```
if diff_method == 'instantaneous':
    ima_j = cube[:, :, 1:]
    ima_j_1 = cube[:, :, 0:-1]
    t_0 = integ_time[0]
    t_j = integ_time[1:]
    t_j_1 = integ_time[0:-1]
    # I think this is right but double check just in case
    inst_diff = (ima_j * (t_j - t_0) - ima_j_1 * (t_j_1 - t_0)) / (t_j - t_j_1)
```

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:126]: Should be renamed `get_std_fullframe_lhs_rhs` since it's just computing standard deviations of cube, not of the medians. Documentation should be changed accordingly.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:177]: Remove space.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:194]: Better to just keep the original variable names and change the ones in the plotting functions directly.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:218]: Remove space.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:270]: Why is this a tuple of one value? Why not just make `sigma_level = 0.5`?

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:1]: Everything is so much more concise and readable. Awesome job!

File: notebooks/ir_ima_visualization/README.md
Comment (FDauphin) [inline notebooks/ir_ima_visualization/README.md:1]: Nice brief explanation 👍🏾

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:253]: Mean or median?

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:272]: Just call `compute_diff_imas`. That way it's already defined before the if-else statement below and it already raises an error in case the method is incorrect.
```
diff = compute_diff_imas(cube, integ_time, diff_method)
```

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:293]: Is there reason for the different plotting routines for cumulative and instantaneous (i.e., plotting `diff` vs `np.abs(diff)`, using different `vmin/vmax`)? If they were the same, the plotting doesn't need to be in an if-else statement.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:295]: Based on previous comment, `diff` is now an array of all the differences. You can delete this line and change all the `diff` to `diff[i]` in the rest of the if statement.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:299]: `np.nanmedian(diff[i])`

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:302]: Same as previous comment: change `diff` to `diff[i]` in the rest of the elif statement.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:306]: `np.nanmedian(diff[i])`. Previously `np.nanmedian(inst_diff)` was getting the median of the entire instantaneous difference cube instead of one element in the instantaneous difference cube.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:77]: Just to be crystal clear, are the ima reads supposed to be subtracted in reverse order? I'm sorry if that was already mentioned in the text but I'm curious as to why for example you would perform SCI[15] - SCI[14] if SCI[14] came _after_ SCI[15] (hence reversed time order) and if for instantaneous it's subtracted in time order (pseudo SCI[14] - SCI[15]).

**Review by FDauphin — CHANGES_REQUESTED** [review]
Great work on the edits; the module and notebook are looking a lot better. Here are some overall comments:

- There are a few things to consider in `panel_plot` that could make it more efficient. Everything else in the module are mostly small changes.
- Make sure to push notebooks with clear outputs. Since this one was pushed with the outputs, a difference file can't be loaded on GitHub, which means I can't add comments directly on here like previously (no worries if you didn't know, just learned that right now haha). I'll add my comments on the notebook in a separate comment.
- Moving forward for PRs, continue editing, committing, and pushing on just _one branch per project_. Since git tracks every change on the branch until it's merged or closed, you can still commit on the same branch during a review, which keeps the reviewing process more organized. By creating a new branch (e.g. IMA_Vis2), when you push, it creates a new PR (e.g. this PR), and it's harder to keep track of what revisions were made from the first PR (e.g. #36). We can chat more about this if you need more clarification.

With all that said, the end of the tunnel is near!

UPDATE: I was able to reopen #43, merge that PR with this PR, and copy over all the comments from #43 to here. Please only commit to `annie_IR_IMA_Vis` for this notebook.

**Comment by FDauphin** [comment]
Here are some comments on the notebooks with which block they appear on:

- Last sentence of first paragraph on "Time-Variable Background in the WFC3/IR Channel"
    - add the parentheses around x<400 (i.e., ... with the leftmost (x<400) columns of the ... )
- Imports
    - `Mast` and `mode` are not being called in the notebook so they can be removed. 
- Section 3. code block 9
    - `path, filename = os.path.split(file)`
- Put all of the functions defined in the notebook in your module.
- `simple_panel_plot`
    - a better name might be `plot_ima_subplots` 
    - Build a simple panel plot of individual IMA reads.
    - Path to a RAW full-frame IR image fits file. (period at end of sentence)
    - Just use `ima_filename` instead of `ima`
- Section 3. code block 13
    - make the `plt.plot` line into two lines for readability
    - e.g., `plt.plot(integ_time[1:], median_fullframe[1:]*integ_time[1:], `
    - `marker = marker_select[i], markersize = 8, color = color_select[i], label = filename)`
- `plot_ramp_subplots`
    - Definition does not match what’s being plot
- `create_panel_plot`
    - a better name might be `plot_ima_difference_subplots`
    - The IMSET difference
    - Path to a RAW full-frame IR fits file.
- About this Notebook
    - don’t forget to change “updated on”

**Comment by FDauphin** [comment]
In the notebook Section 5: WFC3 ISR 2018-05 instead of WFC3/IR ISR 2018-05.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [reply notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:81]: @annierose3 Can you address this real quick?

Comment (annierose3) [reply notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:81]: also, the instantaneous method is psuedo 14-15, but since it is ~derivative-like~ more than just subtraction, it isn't a negative difference in rate and works fine.

**Review by FDauphin — CHANGES_REQUESTED** [review]
Last small comments in the module; the notebook itself is good to go. After those are fixed, we'll be good to merge 👍🏾 

One last thing I wanted to check: The standard deviations for instantaneous method decrease as a function of time, where the standard deviation for the cumulative method increase as a function of time. Is that to be expected?

**Comment by FDauphin** [comment]
> Last small comments in the module; the notebook itself is good to go. After those are fixed, we'll be good to merge 👍🏾
> 
> One last thing I wanted to check: The standard deviations for instantaneous method decrease as a function of time, where the standard deviation for the cumulative method increase as a function of time. Is that to be expected?

Update: The standard deviation for the instantaneous method **_increases_** as a function of time, where the standard deviation for the cumulative method **_decreases_** as a function of time. Is that to be expected?

**Comment by annierose3** [comment]
Checking about the decreasing / increasing standard deviations, right now I'm not sure why that is.

**Comment by FDauphin** [comment]
> Checking about the decreasing / increasing standard deviations, right now I'm not sure why that is. Creating a PR for all other changes!

@annierose3 I spoke with Jen and she said you two also determined the standard deviation issue was because of the overscan regions. Are you going to exclude those regions when calculating median and standard deviations, or just keep everything the same? If you do decide to exclude it, be sure to add a comment in `get_median_fullframe_lhs_rhs` and `get_std_fullframe_lhs_rhs` that you are excluding those regions. In addition (if you decide to exclude those regions), the title in `panel_plot` should be changed accordingly:
```
ax.set_title(f'$\mu = ${median_diff_full_frame[i]:.2f}±{standard_dev_fullframe[i]:.2f} e-/s', fontsize = 30)
```
It's up to you though, everything else is fine and good to merge if you don't decide to proceed with what was stated above.

I also apologize if the last few comments are coming off as nit-picky 😅; just want to make sure your notebook accurately demonstrates what was initially intended

**Comment by annierose3** [comment]
Oops, I thought the changes had been made clipping the boundary pixels but It looks like they I changed them in another version, not this one. Fixing that now!

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:148]: Consider using `std` for variable prefix since it's already being used for function name.

File: notebooks/ir_ima_visualization/ima_visualization_and_differencing.py
Comment (FDauphin) [inline notebooks/ir_ima_visualization/ima_visualization_and_differencing.py:206]: (continuing of refactoring comment)
```
def panel_plot(diff, itime, medians_diff, stds_diff):
    """
    Original function definition with appropriate parameter definitions.
    """
    # previous code in function

    # Unpack variables
    median_diff_fullframe, median_diff_lhs, median_diff_rhs = medians_diff
    standard_dev_fullframe, standard_dev_lhs, standard_dev_rhs = stds_diff

    # rest of code in function
```

**Review by FDauphin — APPROVED** [review]
All my comments were addressed; approving for merge. Exceptional job on this notebook 🙌🏾 
The comments added are "extra credit" if you have time, if not feel free to leave for version 2.0. Please comment if you will complete them or will get back to it another time.

**Review by FDauphin — CHANGES_REQUESTED** [review]
Make sure you commit an empty notebook!

**Comment by annierose3** [comment]
I will commit a cleared notebook now, but I plan to save the rest of these comments for a later version in order to focus on getting the other 2 notebooks ready for publication.

**Review by FDauphin — APPROVED** [review]
Approving for merge.

**Pull request merged**
